OpenCV 4.0.1 + Cuda 10.1, failed to build?

It looks it's still a kind of buggy? Did any of you have successfully built OpenCV 4.0.1 with Cuda 10.1?

```
....../opencv/modules/core/include/opencv2/core/cvstd_wrapper.hpp(45): error: incomplete type is not allowed
          detected during:
            instantiation of "cv::sfinae::has_parenthesis_operator<C, Ret, Args...>::check [with C=cv::DefaultDeleter<<unnamed>::LookUpTableImpl>, Ret=void, Args=<<unnamed>::LookUpTableImpl *>]" based on template argument <cv::DefaultDeleter<<unnamed>::LookUpTableImpl>> 
(49): here
            instantiation of class "cv::sfinae::has_parenthesis_operator<C, Ret, Args...> [with C=cv::DefaultDeleter<<unnamed>::LookUpTableImpl>, Ret=void, Args=<<unnamed>::LookUpTableImpl *>]" 
(138): here
            instantiation of "cv::Ptr<_Tp> cv::makePtr<_Tp,A1...>(const A1 &...) [with _Tp=<unnamed>::LookUpTableImpl, A1=<cv::_InputArray>]" 
....../opencv_contrib/modules/cudaarithm/src/cuda/lut.cu(207): here

....../opencv/modules/core/include/opencv2/core/cvstd_wrapper.hpp(49): error: template instantiation resulted in unexpected function type of "std::true_type (std::integral_constant<__nv_bool, false> *)" (the meaning of a name may have changed since the template declaration -- the type of the template is "std::true_type (std::is_same<std::decay<decltype((<expression>))>::type, void>::type *)")
          detected during:
            instantiation of "cv::sfinae::has_parenthesis_operator<C, Ret, Args...>::check [with C=cv::DefaultDeleter<<unnamed>::LookUpTableImpl>, Ret=void, Args=<<unnamed>::LookUpTableImpl *>]" based on template argument <cv::DefaultDeleter<<unnamed>::LookUpTableImpl>> 
(49): here
            instantiation of class "cv::sfinae::has_parenthesis_operator<C, Ret, Args...> [with C=cv::DefaultDeleter<<unnamed>::LookUpTableImpl>, Ret=void, Args=<<unnamed>::LookUpTableImpl *>]" 
(138): here
            instantiation of "cv::Ptr<_Tp> cv::makePtr<_Tp,A1...>(const A1 &...) [with _Tp=<unnamed>::LookUpTableImpl, A1=<cv::_InputArray>]" 
....../opencv_contrib/modules/cudaarithm/src/cuda/lut.cu(207): here

....../opencv/modules/core/include/opencv2/core/cvstd_wrapper.hpp(52): error: name followed by "::" must be a class or namespace name
          detected during:
            instantiation of class "cv::sfinae::has_parenthesis_operator<C, Ret, Args...> [with C=cv::DefaultDeleter<<unnamed>::LookUpTableImpl>, Ret=void, Args=<<unnamed>::LookUpTableImpl *>]" 
(138): here
            instantiation of "cv::Ptr<_Tp> cv::makePtr<_Tp,A1...>(const A1 &...) [with _Tp=<unnamed>::LookUpTableImpl, A1=<cv::_InputArray>]" 
....../opencv_contrib/modules/cudaarithm/src/cuda/lut.cu(207): here

3 errors detected in the compilation of "/tmp/tmpxft_00006477_00000000-6_lut.cpp1.ii".
-- Removing ....../opencv/build/modules/cudaarithm/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_lut.cu.o
/usr/bin/cmake -E remove ....../opencv/build/modules/cudaarithm/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_lut.cu.o
CMake Error at cuda_compile_1_generated_lut.cu.o.Release.cmake:279 (message):
  Error generating file
  ....../opencv/build/modules/cudaarithm/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_lut.cu.o


modules/cudaarithm/CMakeFiles/opencv_cudaarithm.dir/build.make:6764: recipe for target 'modules/cudaarithm/CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_lut.cu.o' failed
make[2]: *** [modules/cudaarithm/CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_lut.cu.o] Error 1
make[2]: Leaving directory '....../opencv/build'
CMakeFiles/Makefile2:2276: recipe for target 'modules/cudaarithm/CMakeFiles/opencv_cudaarithm.dir/all' failed
make[1]: *** [modules/cudaarithm/CMakeFiles/opencv_cudaarithm.dir/all] Error 2
make[1]: Leaving directory '....../opencv/build'
Makefile:165: recipe for target 'all' failed
make: *** [all] Error 2
```

Cheers

> Summary of build failures on various settings(with CUDA enabled):
> 
> CUDA	OpenCV(Branch)	Compiler	Build Failure
> 10.1	master	GCC 7.3.0	YES (#13941, #13952)
> 10.1	master	GCC 5.4.0	YES (#13941, #13952)
> 10.1	master	VS2015	YES (#13491, #13952)
> 10.1	3.4	GCC 7.3.0	YES (#13941) NO(Fixed)
> 10.1	3.4	GCC 5.4.0	YES (#13941) NO(Fixed)
> 10.1	3.4	VS2015	NO
> 10.0	master	GCC 7.3.0	NO
> 10.0	master	GCC 5.4.0	NO
> 10.0	master	VS2015	YES (#13491) NO(Fixed)
> 10.0	3.4	GCC 7.3.0	NO
> 10.0	3.4	GCC 5.4.0	NO
> 10.0	3.4	VS2015	YES (dynlink_nvcuvid.h not found)


CUDA 10.1 with OpenCV_contrib 3.4 ,GCC 7.3.0	Build error 

imgcodecs_c.h:1:2: error: #error "This header with legacy C API declarations has been removed from OpenCV.

--- a/modules/core/include/opencv2/core/cvstd_wrapper.hpp
+++ b/modules/core/include/opencv2/core/cvstd_wrapper.hpp
@@ -57,10 +57,12 @@ template <typename T, typename = void>
 struct has_custom_delete
         : public std::false_type {};
 
+// Force has_custom_delete to std::false_type when NVCC is compiling CUDA source files
+#ifndef __CUDACC__
 template <typename T>
 struct has_custom_delete<T, typename std::enable_if< sfinae::has_parenthesis_operator<DefaultDeleter<T>, void, T*>::value >::type >
         : public std::true_type {};
-
+#endif
 
 template<typename T>
 struct Ptr : public std::shared_ptr<T>

